beacon-chain/db/kv/migration_state_validators.go

```diff
@@ -0,0 +1,91 @@
+package kv
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/golang/snappy"
+	v1 "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	bolt "go.etcd.io/bbolt"
+)
+
+var migrationStateValidators0Key = []byte("migration_state_validator")
+
+func migrateStateValidators(tx *bolt.Tx) error {
+	if !featureconfig.Get().EnableHistoricalSpaceRepresentation {
+		return nil
+	}
+	mb := tx.Bucket(migrationsBucket)
+	if b := mb.Get(migrationStateValidators0Key); bytes.Equal(b, migrationCompleted) {
+		return nil // Migration already completed.
+	}
+
+	// get the source and destination buckets.
+	log.Infof("migrating %s. It will take few minutes. Please wait.", stateBucket)
+	stateBkt := tx.Bucket(stateBucket)
+	if stateBkt == nil {
+		return nil
+	}
+	valBkt := tx.Bucket(stateValidatorsBucket)
+	if valBkt == nil {
+		return nil
+	}
+	indexBkt := tx.Bucket(blockRootValidatorHashesBucket)
+	if indexBkt == nil {
+		return nil
+	}
+
+	// for each of the state in the stateBucket, do the migration.
+	ctx := context.Background()
+	c := stateBkt.Cursor()
+	for k, v := c.First(); k != nil; k, v = c.Next() {
```

Comment:
Concerned about this mega-transaction. It does a lot and might overwhelm the node's resources in terms of memory and CPU. Bolt really dislikes long-running transactions and this one might be harsh on the node. Worth checking carefully at runtime, or even in a benchmark test that can be cpu and memprofiled with pprof

Reply:
Yes. I have seen memory spiking during migrations. But how do we avoid it. Entire migration call seems to be from inside a single transaction. One good thing is we are deleting many thing as well, this will make sure we use the same old pages instead of allocating new ones inside bolt DB.

Reply:
Current migration semantics restrict us from making it more manageable. Although doing this in one big transaction will most likely cause OOMs for stakers on limited boxes.  We can improve on this in future PRs, as its all behind a flag for now.